### PR TITLE
Merge build systems in the stdlib directory

### PR DESCRIPTION
**Pull request by shindere**
This PR continues the build-systems merging work started in PRs
#762, #764, #785, #788, #811 and #812.

### CODE REVIEW

**Comment by shindere** [comment]
Just rebased on latest trunk. @alainfrisch can you review the code, please?

**Comment by gasche** [comment]
(cc: @dra27, @whitequark (does Adrien Nader have a github account?) who may also be interested in giving an opinion.)

File: stdlib/Makefile [hunk 1]
```diff
@@ -13,66 +13,236 @@
...
+	  strip tmpheader$(EXE) && \
+	  mv tmpheader$(EXE) target_camlheader$$suff; \
+	done && \
+	cp camlheader camlheader_ur
+
+else # Windows
+
+# TODO: see whether there is a way to further merge the rules below
+# with those above
+
```
Comment (dra27) [inline stdlib/Makefile:156]: Only with tremendous care - I think more than one of the lines below has been the subject of its own Mantis PR!

**Comment by dra27** [comment]
Have you compilation with flexdll built using the submodule (rather than preinstalled)? I of course don't mind doing it, though from a testing perspective, I'd prefer to review/test all these merges at once, even if they're in separate PRs...

**Comment by shindere** [comment]
David Allsopp (2016/11/09 08:00 -0800):

> Have you tested flexdll built using the submodule?

No, not yet. Am I correct that you are talking about boostrapping
flexdll? In that case: is there already a documentation explaining how
this can be tested? As far as I remember it was not obvious how to do it
last time I tried. Moreover flexdll is already installed on the machines
where I do the Windows tests.

@dra27 if you can test the build with the flexdll submodule I'd really
appreciate. This PR is actually the only unmerged one at the moment,
regarding the merge of build systems.

**Comment by dra27** [comment]
No problem - I will try to this evening. Assuming you're using a git clone: `git submodule update --init` and then `make -f Makefile.nt flexdll` (you can include world/world.opt after flexdll). It should ignore the system-installed flexlink - though it's a good idea not to have one while testing.

**Comment by whitequark** [comment]
My opinion is that this PR set is rather wonderful and is certain to make my life (as a cross-compilation package maintainer) rather easier, even if there are some omissions.

**Comment by shindere** [comment]
Thanks, David, for your prompt responses and your support!

David Allsopp (2016/11/09 08:21 -0800):

> No problem - I will try to this evening.

Nice! Thanks!

> Assuming you're using a git clone:

Yes, I do.

> `git submodule update --init` and then `make -f Makefile.nt flexdll`
> (you can include world/world.opt after flexdll).

Okay I'll try this tomorrow, thanks!

> It should ignore the system-installed flexlink - though it's a good
> idea not to have one while testing.

Given that I do my tests on the CI machines, not having flexdll
installed is not really an option. Well it oculd be, if there is the
choice to also test flexdll, but at the moment it's not the case.

What I can do, though, is to remove flexdll from the PATH.

And I'll try this tomorrow.

**Comment by dra27** [comment]
Removing from PATH is safely equivalent to not having it - there's nothing in the build system which ever try to search for it.

**Comment by dra27** [comment]
Just tried it with msvc x86 and bootstrapped flexdll and everything seems fine!

File: stdlib/Makefile [hunk 2]
```diff
@@ -13,66 +13,236 @@
...
+  camlheaderd target_camlheaderd \
+  camlheaderi target_camlheaderi
+
+ifeq "$(HASHBANGSCRIPTS)" "true"
+$(CAMLHEADERS): ../config/Makefile
+	for suff in '' d i; do \
+	  echo '#!$(BINDIR)/ocamlrun'$$suff > camlheader$$suff && \
+	  echo '#!$(TARGET_BINDIR)/ocamlrun'$$suff >target_camlheader$$suff; \
+	done && \
+	echo '#!' | tr -d '\012' > camlheader_ur;
```
Comment (adrien-n) [inline stdlib/Makefile:131]: This is most probably for another patchset:

```
printf "#!" > camlheader_ur
```

would be much cleaner. I'm not sure we can get into this branch on platforms with \r\n end-of-lines however.

**Comment by gasche** [comment]
@dra27 Thanks for the testing! Feel free to merge if you think this is good.

**Review by adrien-n — APPROVED** [review]
Nothing obviously wrong stood out. The diff seems more difficult to read than it ought to be because git doesn't seem to notice this is a move but I assume you've simply moved the content of Makefile.nt around and wrapped it in a conditional.

**Pull request merged**